src/RequestFactory.php

```diff
@@ -246,4 +302,73 @@ private function isJson(array $headers)
         // http://greenbytes.de/tech/webdav/rfc2616.html#rfc.section.14.17
         return strpos($headers['Content-Type'], 'application/json') === 0;
     }
+
+    /**
+     * @param array $headers
+     */
+    private function headerIsJsonOrFail(array $headers) {
+        if (!$this->isJson($headers)) {
+            throw new  RuntimeException("Content-Type header is not equal to \'application/json'");
+        }
+    }
+
+    /**
+     * @param string $uri
+     * @param array  $request
+     * @return array
+     */
+    private function getPlaceholdersFromUriAndHeader($uri, $request) {
+        $result = [];
+        $placeholders = $this->getPlaceholdersInString($uri);
+        $placeholders[] = $this->getPlaceholdersInArray($request['headers']);
+
+        array_walk_recursive($placeholders,function($v, $k) use (&$result){ $result[] = $v; });
+
+        return $result;
+    }
+
+    /**
+     * @param array $placeholders
+     * @param \JsonSerializable $jsonSerializable
+     * @return array
+     */
+    private function getPlaceholderValuesFromJson(array $placeholders = [], \JsonSerializable $jsonSerializable) {
+        $json = \GuzzleHttp\json_encode($jsonSerializable);
+        $jsonAsArray = \GuzzleHttp\json_decode($json, true);
+        $values = [];
+
+        foreach ($placeholders as $placeholder) {
+            $keysArray = explode('.', $placeholder);
+
+            if (!$keysArray) {
+                $keysArray = [$placeholder];
+            }
+
+            $currentArray = $jsonAsArray;
+            $currentValue = [];
+
+            foreach ($keysArray as $key) {
```

Comment:
I can't follow this logic, could you encapsulate it to some functions that are more meaningful?